internal/clearingway/menu.go

```diff
@@ -0,0 +1,166 @@
+}
+
+type Menu struct {
+	Name           string  // internal name e.g "menureclear"
+	Type           MenuType
+	Blurb          string
+	Title          string  // title to show in embed
+	Description    string  // optional description to show in embed
+	ImageURL       string  // optional image URL
+	AdditionalData MenuComponent  // additional data depending on MenuType
+}
+
+type MenuComponent interface{
+	Type() MenuType
+}
+
+type MenuRoleHelper struct {
+	Role         *Role
+	Prerequisite *Role
+}
+
+type MenuTypeEncounterData struct {
+	Roles        map[string]*MenuRoleHelper
+	ExtraRoles   []*Role
+	RoleType     []RoleType
+	MultiSelect  bool
+	RequireClear bool
+}
+
+func (MenuTypeEncounterData) Type() MenuType {
+	return MenuTypeEncounter
+}
+
+func (m *Menu) Init(c *ConfigMenu) {
+	m.Name = c.Name
+	m.Type = MenuType(c.Type)
+	m.Blurb = c.Blurb
+	m.Title = c.Title
+
+	if len(c.Description) != 0 {
+		m.Description = c.Description
+	}
+
+	if len(c.ImageUrl) != 0 {
+		m.ImageURL = c.ImageUrl
+	}
+
+	switch m.Type {
+	case MenuTypeDefault:
+	case MenuTypeEncounter:
+		m.Type = MenuTypeEncounter
+		m.AdditionalData = &MenuTypeEncounterData{}
+		data := m.AdditionalData.(*MenuTypeEncounterData)
+
+		if len(c.RoleType) != 0 {
+			for _, roleType := range c.RoleType {
+				data.RoleType = append(data.RoleType, RoleType(roleType))
+			}
+		}
+
+		if c.MultiSelect {
+			data.MultiSelect = true
+		} else {
+			data.MultiSelect = false
+		}
+
+		if c.RequireClear {
+			data.RequireClear = true
+		} else {
+			data.RequireClear = false
+		}
+
+		for _, configRole := range c.ConfigRoles {
+			role := &Role{}
+			if len(configRole.Name) != 0 {
+				role.Name = configRole.Name
+			}
+			if len(configRole.Description) != 0 {
+				role.Description = configRole.Description
+			}
+			if configRole.Color != 0 {
+				role.Color = configRole.Color
+			}
+			if configRole.Hoist {
+				role.Hoist = true
+			}
+			if configRole.Mention {
+				role.Mention = true
+			}
+			data.ExtraRoles = append(data.ExtraRoles, role)
+		}
+	}
+}
+
+func (g *Guild) DefaultMenus() {
+	g.Menus.Defaults[MenuMain] = &Menu{
+		Name: MenuMain,
+		Type: MenuTypeDefault,
+		Title: "Welcome to " + g.Name,
+		Description: "Use the buttons below to assign roles!",
+	}
+
+	g.Menus.Defaults[MenuVerify] = &Menu{
+		Name: MenuVerify,
+		Type: MenuTypeDefault,
+		Title: "Verify Character",
+	}
+
+	g.Menus.Defaults[MenuRemove] = &Menu{
+		Name: MenuRemove,
+		Type: MenuTypeDefault,
+		Title: "Remove Roles",
+		Description: "Use the buttons below to remove Clearingway related roles!",
+	}
+}
+
+// ComponentsHandler is function that executes the respective guild specific functions
+func (c *Clearingway) ComponentsHandler(s *discordgo.Session, i *discordgo.InteractionCreate, customID string){
+	if g, ok := c.Guilds.Guilds[i.GuildID]; ok {
+		g.ComponentsHandlers[customID](s, i)
+	} else {
+		fmt.Printf("Interaction received from guild %s with no configuration!\n", i.GuildID)
+		return
+	}
+}
```

Comment:
this will probably cause us pain in the future when we start implementing this for other servers. not necessarily a change suggestion but we should probably plan out more of how we're going to be dealing with the different variations of roles/criteria to get those roles. one option is a setup menu and parse in all of these variations before displaying the actual menu. this is pretty big overhead work but may pay off later. feel free to actually draw a diagram of how exactly you want things to be handled first before actually coding it

Reply:
I've been sitting on this and trying to think of a better way to do this for the past couple days but I'm drawing a blank on how to do this differently. I've tried to seek alternatives and mocked up something similar to what already exists (`InteractionCreate()` in `handlers.go`). However, a problem this approach introduced the need to differentiate between what command to execute. Since each button in the menu has different purposes (e.g sending the verify modal, role assignment, role removal, sending a submenu) this would result in a single long switch case since code between each button is not really shared. To help this issue, I tried to split up the interactions into categories. This approach resulted in a nest of switch case statements, which I believe is pretty confusing.
I'm struggling really hard to put what I want to say in words (took me an hour just to type this out) but a quick look at the other PRs (#33 and #34) might also help. If it doesn't I could try to maybe explain it in a call?

Reply:
in situation like these, case statements are fine and typically used in command line interfaces (cli). our menu is basically acting as a cli but in gui form. if you want to see an example of awful case nesting, refer to getopt in rpcapd (short for "remote packet capture daemon") https://github.com/rpcapd-linux/rpcapd-linux/blob/fcbb0680d2130d37dd13b4746f7c3d0341a25070/rpcapd.c#L153

getopt is basically a function to read command line arguments used typically in c/c++ https://www.geeksforgeeks.org/getopt-function-in-c-to-parse-command-line-arguments